lib/options.c

```diff
@@ -40,371 +40,249 @@
+ * the socket TCTI).
  */
-tcti_map_entry_t tcti_map_table[] = {
-#ifdef HAVE_TCTI_DEV
-    {
-        .name = "device",
-        .type = DEVICE_TCTI,
-    },
-#endif
-#ifdef HAVE_TCTI_SOCK
-    {
-        .name = "socket",
-        .type = SOCKET_TCTI,
-    },
-#endif
 #ifdef HAVE_TCTI_TABRMD
-    {
-        .name = "tabrmd",
-        .type = TABRMD_TCTI,
-    },
+  #define TCTI_DEFAULT_STR  "abrmd"
+#elif HAVE_TCTI_SOCK
+  #define TCTI_DEFAULT_STR  "socket"
+#elif  HAVE_TCTI_DEV
+  #define TCTI_DEFAULT_STR  "device"
 #endif
-    {
-        .name = "unknown",
-        .type = UNKNOWN_TCTI,
-    },
-};
-/*
- * Convert from a string to an element in the TCTI_TYPE enumeration.
- * An unkonwn name / string will map to UNKNOWN_TCTI.
- */
-TCTI_TYPE
-tcti_type_from_name (char const *tcti_str)
-{
-    int i;
 
-    if (tcti_str == NULL)
-        goto out;
-    for (i = 0; i < N_TCTI; ++i)
-        if (strcmp (tcti_str, tcti_map_table[i].name) == 0)
-            return tcti_map_table[i].type;
-out:
-    return UNKNOWN_TCTI;
-}
-/*
- * Convert from an element in the TCTI_TYPE enumeration to a string
- * representation.
- */
-char *
-tcti_name_from_type (TCTI_TYPE tcti_type)
-{
-    int i;
-    for (i = 0; i < N_TCTI; ++i)
-        if (tcti_type == tcti_map_table[i].type)
-            return tcti_map_table[i].name;
-    return NULL;
-}
-/*
- * Test a common_opts_t structure to be sure the member data has been
- * populated. We don't do any tests on the data for appropriate formats
- * (like testing socket_address for a valid IP address).
- * return 0 if sanity test passes
- * return 1 if help message was explicitly requested
- * return 2 if sanity test fails
- */
-int
-sanity_check_common (common_opts_t  *opts)
-{
-    if (opts->help)
-        return 1;
-    switch (opts->tcti_type) {
-#ifdef HAVE_TCTI_DEV
-    case DEVICE_TCTI:
-        if (opts->device_file == NULL) {
-            fprintf (stderr, "missing --device-file, see --help\n");
-            return 2;
-        }
-        break;
+#ifdef HAVE_TCTI_TABRMD
+#include "tpm2_tools_tcti_abrmd.h"
 #endif
 #ifdef HAVE_TCTI_SOCK
-    case SOCKET_TCTI:
-        if (opts->socket_address == NULL) {
-            fprintf (stderr, "missing --socket-address, see --help\n");
-            return 2;
-        }
-        if (opts->socket_port == 0) {
-            fprintf (stderr, "missing --socket-port, see --help\n");
-            return 2;
-        }
-        break;
+#include "tpm2_tools_tcti_socket.h"
 #endif
-#ifdef HAVE_TCTI_TABRMD
-    case TABRMD_TCTI:
-        /* no options for this one yet */
-        break;
+#ifdef HAVE_TCTI_DEV
+#include "tpm2_tools_tcti_device.h"
 #endif
-    default:
-        fprintf (stderr, "invalid TCTI, see --help\n");
-        return 2;
+
+/* Environment variables usable as alternatives to command line options */
+#define TPM2TOOLS_ENV_TCTI_NAME      "TPM2TOOLS_TCTI_NAME"
+
+struct tpm2_options {
+    tpm2_option_handler on_opt;
+    tpm2_arg_handler on_arg;
```

Comment:
I think that calling these {opt,arg}_callback or {opt,arg}_handler is much more clear.